lib/DefinePlugin.js

```diff
@@ -2,129 +2,121 @@
+		let definitions = this.definitions;
+		compiler.plugin("compilation", (compilation, params) => {
+			compilation.dependencyFactories.set(ConstDependency, new NullFactory());
+			compilation.dependencyTemplates.set(ConstDependency, new ConstDependency.Template());
 
-			function stringifyObj(obj) {
-				return "{" + Object.keys(obj).map(function(key) {
-					var code = obj[key];
-					return JSON.stringify(key) + ":" + toCode(code);
-				}).join(",") + "}";
-			}
+			params.normalModuleFactory.plugin("parser", (parser) => {
+				(function walkDefinitions(definitions, prefix) {
+					Object.keys(definitions).forEach((key) => {
+						let code = definitions[key];
+						if(code && typeof code === "object" && !(code instanceof RegExp)) {
+							walkDefinitions(code, prefix + key + ".");
+							applyObjectDefine(prefix + key, code);
+							return;
+						}
+						applyDefineKey(prefix, key);
+						applyDefine(prefix + key, code);
+					});
+				}(definitions, ""));
 
-			function toCode(code) {
-				if(code === null) return "null";
-				else if(code === undefined) return "undefined";
-				else if(code instanceof RegExp && code.toString) return code.toString();
-				else if(typeof code === "function" && code.toString) return "(" + code.toString() + ")";
-				else if(typeof code === "object") return stringifyObj(code);
-				else return code + "";
-			}
+				function stringifyObj(obj) {
+					return "{" + Object.keys(obj).map((key) => {
+						let code = obj[key];
+						return JSON.stringify(key) + ":" + toCode(code);
+					}).join(",") + "}";
+				}
 
-			function applyDefineKey(prefix, key) {
-				var splittedKey = key.split(".");
-				splittedKey.slice(1).forEach(function(_, i) {
-					var fullKey = prefix + splittedKey.slice(0, i + 1).join(".");
-					parser.plugin("can-rename " + fullKey, function() {
-						return true;
-					});
-				});
-			}
+				function toCode(code) {
+					if(code === null) return "null";
+					else if(code === undefined) return "undefined";
+					else if(code instanceof RegExp && code.toString) return code.toString();
+					else if(typeof code === "function" && code.toString) return "(" + code.toString() + ")";
+					else if(typeof code === "object") return stringifyObj(code);
+					else return code + "";
+				}
 
-			function applyDefine(key, code) {
-				var isTypeof = /^typeof\s+/.test(key);
-				if(isTypeof) key = key.replace(/^typeof\s+/, "");
-				var recurse = false;
-				var recurseTypeof = false;
-				code = toCode(code);
-				if(!isTypeof) {
-					parser.plugin("can-rename " + key, function() {
-						return true;
+				function applyDefineKey(prefix, key) {
+					const splittedKey = key.split(".");
+					splittedKey.slice(1).forEach((_, i) => {
+						const fullKey = prefix + splittedKey.slice(0, i + 1).join(".");
+						parser.plugin("can-rename " + fullKey, () => true);
 					});
-					parser.plugin("evaluate Identifier " + key, function(expr) {
-						if(recurse) return;
-						recurse = true;
```

Comment:
Aye! Will add this back. What could this lead this to? So i can check if there are open issues related to this/write a test for it